File: study2_200211/jiho_LRU_Caching.js
const getKeyOfMinValue = cache => {
  extractedKey = undefined;
  minValue = Infinity;
  for (let key in cache) {
    newValue = cache[key];
    if (newValue < minValue) {
      extractedKey = key;
      minValue = newValue;
    }
  }
  return extractedKey;
};

function solution(cacheSize, cities) {
  let cost = 0;
  cache = {};
  // O(N)
  for (let [priority, city] of cities.entries()) {
    city = city.toLowerCase();
    if (city in cache) {
      cost += 1;
      cache[city] = priority + 1;
    } else {
      if (Object.keys(cache).length === cacheSize) {
        // O(30)
        extractedKey = getKeyOfMinValue(cache);
        delete cache[extractedKey];
      }
      cache[city] = priority;
      cost += 5;
    }
  }
  return cost;
}

console.log(solution(1, ["jeju", "Jeju", "Jeju"]));

File: study2_200211/jiho_LRU_Caching.py
import heapq


def solution(cache_size, datas):
    cost = 0
    cache = dict()
    # O(N)
    for idx, data in enumerate(datas):
        if cache.get(data, False):
            cost += 1
            cache[data] = idx + 1
        else:
            if len(cache) == cache_size:
                extracted_key = None
                min_value = float("inf")
                # O(30)
                for key, value in cache.items():
                    if value < min_value:
                        extracted_key = key
                        min_value = value
                del cache[extracted_key]
            cache[data] = idx + 1
            cost += 5
    return cost


print(
    solution(
        3,
        [
            "Jeju",
            "Pangyo",
            "Seoul",
            "NewYork",
            "LA",
            "Jeju",
            "Pangyo",
            "Seoul",
            "NewYork",
            "LA",
        ],
    )
)
print(
    solution(
        3,
        [
            "Jeju",
            "Pangyo",
            "Seoul",
            "Jeju",
            "Pangyo",
            "Seoul",
            "Jeju",
            "Pangyo",
            "Seoul",
        ],
    )
)
print(
    solution(
        5,
        [
            "Jeju",
            "Pangyo",
            "Seoul",
            "NewYork",
            "LA",
            "SanFrancisco",
            "Seoul",
            "Rome",
            "Paris",
            "Jeju",
            "NewYork",
            "Rome",
        ],
    )
)
